### PR TITLE
Add the ability to reload the iframe

### PR DESCRIPTION
**Pull request by pento**
After logging in, cookies are set, but the iframe needs to be reloaded to recognise them properly. This PR exports a `reloadProxy()` function for doing that. 

Unfortunately, adding the new export has the side effect of breaking anything that was previously importing `wpcom-proxy-request` as `require( 'wpcom-proxy-request' )`.

This PR also fixes a couple of race conditions:
- `uninstall()` didn't set `loaded` to `false`, so `request()` could try to send the request before the iframe was reloaded.
- `onload()` was hooked to the iframe's `load` event, which fires when the iframe is inserted into the DOM, before the iframe is ready to receive messages. It now waits for a "ready" message from the iframe (see D6171 for the corresponding iframe code).

The changes to `package.json` exist to allow `wpcom-proxy-request` branches to be installed as dependencies. 

See Automattic/wp-calypso#15543.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,7 +21,9 @@
     "Andy Skelton <andy@automattic.com>"
   ],
   "files": [
+    "index.js",
```
Comment (yurynix) [inline package.json]: Why is this change required?

Comment (yurynix) [inline package.json]: I mean, is it just for testing or that's a permanent change?

Comment (pento) [inline package.json]: It's there for installing the repo as a dependency - I'm not sure if there's a better way to do it, but `npm` seems to download a copy of the repo, then delete everything not in the `files` list. So, I added `index.js` and `Makefile`, to ensure they'd be there.

If there's an option for running `make build` before `npm` removes the files, that would be better.

File: index.js [hunk 2]
```diff
@@ -401,6 +408,13 @@ function onmessage( e ) {
 		data = JSON.parse( data );
 	}
 
+	// Once the iframe is loaded, we can start using it.
+	// The proxy iframe defaults to sending as a string, so test both string and array.
+	if ( 'ready' === data[0] || '["ready",null,null]' === data ) {
```
Comment (pento) [inline index.js]: This logic is only necessary for IE8-9 and Firefox <= 6, per [this comment](https://github.com/Automattic/wpcom-proxy-request/blob/master/index.js#L26-L34). Given that Calypso doesn't work with them, and even Core has [dropped support for them](https://make.wordpress.org/core/2017/04/23/target-browser-coverage/), I'm inclined to assume the browser supports object transfers.

(A follow up PR can remove the logic here and in the iframe for the existing handling of old browsers.)

File: package.json (hunk 1 above)
Comment (yurynix) [inline package.json]: I think the "normal flow" is shipping to `npm` the built version, so we only supped to ship the bundle imo

Comment (pento) [inline package.json]: Yah, I was kind of hoping for a way to not have to change the `files` list. If there isn't, I'll just revert the changes to `package.json` before I merge.

**Comment by yurynix** [comment]
Looks good 👍 

> Unfortunately, adding the new export has the side effect of breaking anything that was previously importing wpcom-proxy-request as require( 'wpcom-proxy-request' ).

Maybe we should mention that in the README.md ?

File: index.js [hunk 3]
```diff
@@ -335,9 +335,6 @@ function install() {
 	// create the <iframe>
 	iframe = document.createElement( 'iframe' );
 
-	// set `loaded` to true once the "load" event happens
-	event.bind( iframe, 'load', onload );
```
Comment (Tug) [inline index.js]: I think we should keep this as well so we're backward compatible with the current version of the rest-proxy. We just need to run onload once either from a `load` event on the iframe or from a `ready` message right?

File: package.json [hunk 4]
```diff
@@ -10,6 +10,7 @@
     "cookie"
   ],
   "scripts": {
+    "postinstall": "postinstall-build build \"make build\"",
```
Comment (Tug) [inline package.json]: I'm not sure why this is needed, the `prepublish` hook is run on `install` (when installed as a dependency).

**Comment by Tug** [comment]
LGTM 👍 

I left a couple of comments.

> onload() was hooked to the iframe's load event, which fires when the iframe is inserted into the DOM, before the iframe is ready to receive messages. It now waits for a "ready" message from the iframe (see D6171 for the corresponding iframe code).

`onload` is supposed to be triggered when the iframe has fully loaded so any synchronous javascript defined there should have had time to execute.
I think the reason why we had some errors on calypso is because of the user bootstrap request which is emitted on calypso boot, before the dom is even ready and the iframe even inserted (we had this problem on wp-desktop at least).

In any case, we should not forget to deploy the server patch first :)

File: index.js (hunk 3 above)
Comment (pento) [inline index.js]: Now that D6171 has landed, the iframe will always send the `ready` message - we don't support having the proxy iframe run elsewhere, so it doesn't seem necessary to maintain backwards compatibility here.

File: package.json (hunk 4 above)
Comment (pento) [inline package.json]: I'm just going to revert the change to `package.json` before merging. It's pretty hacky, and not necessary for release.

**Comment by pento** [comment]
> onload is supposed to be triggered when the iframe has fully loaded so any synchronous javascript defined there should have had time to execute.

Yeah, it's supposed to work like that, but I was continually running into a problem where the `load` event was fired, but the iframe had not yet hooked into the `message` event, which should be happening on initial parse, before the `load` event fires.

Before merge, it seems like the only thing left to do is to revert the hacks in `package.json`.

After merge, we'll need to bump the version to `4.0.0`, as this is a breaking change, and add an entry to `history.md`:

> Add the exported `reloadProxy()` function, for reloading the proxy iframe. This change is not backwards compatible with importing the default function using `require()`, as we're now exporting an object, instead of a function. `const proxy = require( 'wpcom-proxy-request' );` will need to be changed to `const proxy = require( 'wpcom-proxy-request' ).default;`. `import proxy from 'wpcom-proxy-request';` will continue to work correctly.

I haven't done an NPM package release before, so I don't know what else needs to happen. 🙂

**Comment by yurynix** [comment]
That's good to go IMO 🚢

**Comment by Tug** [comment]
LGTM 👍
